From e098164a49f1b3ed99c7382dc4a303e5a44ec850 Mon Sep 17 00:00:00 2001
From: GUO Zihua <guozihua@huawei.com>
Date: Tue, 19 Jul 2022 22:10:08 +0800
Subject: [PATCH 3609/4585] [Backport] KEYS: Fix mistaken sizeof call in
 pgp_key_generate_id

hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I5H61R

Reference: https://gitee.com/openeuler/kernel/commit/886e6e2458ccd70aee6b17f5b40638cde92aa89e

--------------------------------

pgp_key_generate_id() is trying get the size of a flexible length
structure, however the sizeof() is called on the pointer itself.
Besides, considering it's trying to get the size of a flexible length
structure, use struct_size() instead.

Fixes: 4006f47d4e21 ("KEYS: PGP data parser")
Signed-off-by: GUO Zihua <guozihua@huawei.com>
Reviewed-by: Roberto Sassu <roberto.sassu@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: GUO Zihua <guozihua@huawei.com>
---
 crypto/asymmetric_keys/pgp_public_key.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/crypto/asymmetric_keys/pgp_public_key.c b/crypto/asymmetric_keys/pgp_public_key.c
index 98b1707a0164..928029a13435 100644
--- a/crypto/asymmetric_keys/pgp_public_key.c
+++ b/crypto/asymmetric_keys/pgp_public_key.c
@@ -283,7 +283,8 @@ static struct asymmetric_key_ids *pgp_key_generate_id(
 		goto error;
 
 	kids->id[0] = kid;
-	kids->id[1] = kmemdup(kid, sizeof(kid) + fingerprint_len, GFP_KERNEL);
+	kids->id[1] = kmemdup(kid, struct_size(kid, data, fingerprint_len),
+			      GFP_KERNEL);
 	if (!kids->id[1])
 		goto error;
 
-- 
2.17.1

